google/cloud/spanner/timestamp.h

```diff
@@ -16,26 +16,137 @@
- * from the clock's epoch.
+ * A representation of the Spanner TIMESTAMP type: An instant in time.
  *
- * Most implementations use an epoch of 1970-01-01T00:00:00+00:00, where the
- * range of values would be about 1677-09-21 through to 2262-04-11.
+ * A `Timestamp` represents an absolute point in time (i.e., is independent of
+ * any time zone), with at least nanosecond precision, and with a range of at
+ * least 0001-01-01T00:00:00Z to 9999-12-31T23:59:59.999999999Z.
  *
- * Note that this may differ from std::chrono::system_clock::time_point on a
- * particular platform.
+ * Note that the string and protobuf representations share those precision and
+ * range guarantees.
+ *
+ * The `std::chrono::time_point` conversions depend on the precision/range of
+ * the underlying duration and on the `std::chrono::system_clock` epoch.
  */
-using Timestamp = std::chrono::time_point<std::chrono::system_clock,
-                                          std::chrono::nanoseconds>;
+class Timestamp {
+ private:
+  // Convenience alias. `std::chrono::sys_time` since C++20.
+  template <class Duration>
+  using sys_time = std::chrono::time_point<std::chrono::system_clock, Duration>;
+
+ public:
+  /// Default construction yields the Unix epoch.
+  Timestamp() : Timestamp(0, 0) {}
+
+  /// Conversion from/to RFC3339 string.
+  ///@{
+  static StatusOr<Timestamp> FromString(std::string const&);
+  std::string ToString() const;
+  ///@}
+
+  /// Conversion from/to `google::protobuf::Timestamp`. These conversions never
+  /// fail, but may accept/produce protobufs outside their documented range.
+  ///@{
+  static Timestamp FromProto(google::protobuf::Timestamp const&);
+  google::protobuf::Timestamp ToProto() const;
+  ///@}
+
+  /// Conversion from/to `std::chrono::time_point` on the system clock.
+  /// May produce out-of-range errors in either direction, depending on the
+  /// properties of `Duration` and the `std::chrono::system_clock` epoch.
+  ///@{
+  template <class Duration>
+  static StatusOr<Timestamp> FromChrono(sys_time<Duration> const& tp) {
+    return FromRatio((tp - UnixEpoch<Duration>()).count(),
+                     Duration::period::num, Duration::period::den);
+  }
+  template <class Duration>
+  StatusOr<sys_time<Duration>> ToChrono() const {
+    auto const count =
+        ToRatio(std::numeric_limits<typename Duration::rep>::min(),
+                std::numeric_limits<typename Duration::rep>::max(),
+                Duration::period::num, Duration::period::den);
+    if (!count) return count.status();
+    return UnixEpoch<Duration>() +
+           Duration(static_cast<typename Duration::rep>(*count));
+  }
+  ///@}
+
+  /// Build a `Timestamp` from raw counts of seconds/nanoseconds since the Unix
+  /// epoch (minus leap seconds), like a C++17 `std::timespec` but out-of-range
+  /// `nsec` values are normalized. Fails if the time cannot be represented.
+  static StatusOr<Timestamp> FromCounts(std::intmax_t sec, std::intmax_t nsec);
```

Comment:
I'd imagine users should never need this function. Could we move this to a non-member in an internal namespace?

Reply:
I'm not so sure that users will never need it.  It is hard to predict how they might originate their time values.  We certainly make good use of it in our tests.

That said, it is a little strange that it doesn't have a "To" partner.  If I follow through on the "`ToProto()` is the general accessor" idea, then perhaps `FromProto()` is the general factory instead.  But my problem with that is that I wanted a normalizing sec/nsec factory, and flagrantly violating the range documentation of `google::protobuf::Timestamp::nanos` didn't feel right.